### PR TITLE
Add scripts for automatically updating the version of the PowerShell extension

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

Adds scripts for:

- Updating the PowerShell extension's entry in the AzureDataStudio repo
- Updating the PowerShell extension's own version (allows incrementing or setting a new version)
- WIP: Update PSES's version

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [ ] PR has tests
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

**Comment by rjmholt** [comment]
Example ADS PR: https://github.com/Microsoft/azuredatastudio/pull/5247

Example PSES PR: https://github.com/rjmholt/PowerShellEditorServices/pull/4

Example vscode-PowerShell PR: https://github.com/rjmholt/vscode-powershell/pull/6

**Comment by rjmholt** [comment]
One thing I'm ambivalent about is the PSES release script.

It more suitably belongs in the PSES repo, but it uses the modules in this repo. In the PowerShell repo there are build steps that download GitHub assets and it makes the build very hard to follow, reason about, debug or change and I wanted to avoid that. But philosophically PSES is its own thing.

Since we currently always release both together, I can see justification for keeping all of it in vscode-PowerShell for now, especially since it simplifies things. But I don't entirely like it either.

**Comment by TylerLeonhardt** [comment]
What's the reasoning for the version update scripts? So we just need to run that instead of manually sending a PR to rev the versions?

File: tools/repoUpdateScripts/updatePsesVersions.ps1 [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+$prParams = @{
+    Branch = $BranchName
+    Title = "Update PowerShellEditorServices version to $NewVersion"
+    GitHubToken = $GitHubToken
+    Organization = $TargetFork
+    Repository = 'PowerShellEditorServices'
+    Description = $PRDescription
+    FromOrg = 'rjmholt'
+}
+OpenGitHubPr @prParams
```
Comment (TylerLeonhardt) [inline tools/repoUpdateScripts/updatePsesVersions.ps1]:
```suggestion
OpenGitHubPr @prParams

```

File: tools/repoUpdateScripts/updateAzureDataStudio.ps1 [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    # Create the new file contents with the inserted segment
+    $newGalleryFileContent = ReplaceStringSegment -String $galleryFileContent -NewSegment $entryStr -StartIndex $startOffset -EndIndex ($endOffset+1) -AutoIndent
+
+    # Write out the new entry
+    SetFileContent $GalleryFilePath $newGalleryFileContent
+}
+
+$repoLocation = Join-Path ([System.IO.Path]::GetTempPath()) 'ads-temp-checkout'
+
+if (-not $BranchName)
```
Comment (TylerLeonhardt) [inline tools/repoUpdateScripts/updateAzureDataStudio.ps1]: Can you put this in the default parameters?

File: tools/repoUpdateScripts/updateAzureDataStudio.ps1 [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+{
+    $BranchName = "update-psext-$ExtensionVersion"
+}
+
+if (-not $PRDescription)
```
Comment (TylerLeonhardt) [inline tools/repoUpdateScripts/updateAzureDataStudio.ps1]: Can you put this in the default parameters?

File: tools/GitHubTools.psm1 [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+function CloneRepo
+{
+    param(
+        [Parameter(Mandatory)]
+        [string]
+        $OriginRemote,
+
+        [Parameter(Mandatory)]
+        [string]
+        $Destination,
```
Comment (TylerLeonhardt) [inline tools/GitHubTools.psm1]: maybe this could just have a default of `""`

File: tools/repoUpdateScripts/updateAzureDataStudio.ps1 (hunk 2 above)
Comment (rjmholt) [reply tools/repoUpdateScripts/updateAzureDataStudio.ps1]: It depends on another parameter. Does that work?

File: tools/repoUpdateScripts/updateAzureDataStudio.ps1 (hunk 3 above)
Comment (rjmholt) [reply tools/repoUpdateScripts/updateAzureDataStudio.ps1]: Ditto above

File: tools/repoUpdateScripts/updateAzureDataStudio.ps1 (hunk 2 above)
Comment (rjmholt) [reply tools/repoUpdateScripts/updateAzureDataStudio.ps1]: Ah, experimenting shows this depends on the order of parameters

**Comment by rjmholt** [comment]
> What's the reasoning for the version update scripts? So we just need to run that instead of manually sending a PR to rev the versions after a release?

Yeah, just another component in being able to press the release button and having everything happen that needs to happen; automating the mindless without loss of oversight.

File: tools/GitHubTools.psm1 (hunk 4 above)
Comment (rjmholt) [reply tools/GitHubTools.psm1]: `""` would just error I think.

I tried to get a sensible default here, which would be the repo name, but it turns out [that can be complicated](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt-ltdirectorygt)

Comment (rjmholt) [reply tools/GitHubTools.psm1]: So I decided to leave it as mandatory for now, let the human figure it out

File: tools/FileUpdateTools.psm1 [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        }
+    }
+}
+
+function GetVersionFromSemVer
+{
+    [OutputType([version])]
+    param(
+        [Parameter(Mandatory)]
+        [semver]
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply tools/FileUpdateTools.psm1]: No, only PS 6. There's no simple way to manage it in lower versions, so I'm going to try running the tasks in PS 6 only first

File: tools/FileUpdateTools.psm1 [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    finally
+    {
+        $jsonWriter.Dispose()
+        $stringWriter.Dispose()
+    }
+    return $stringBuilder.ToString().Replace([System.Environment]::NewLine, "`r`n")
+}
+
+function SetFileContent
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply tools/FileUpdateTools.psm1]: Was initially to work around the UTF8 BOM thing, but I think you're right that Set-Content is best

File: tools/repoUpdateScripts/updateExtensionVersions.ps1 [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+function FindRequiredPsesVersionSpan
+{
+    param(
+        [Parameter(Mandatory)]
+        [string]
+        $MainTsContent
+    )
+
+    $pattern = [regex]'const\s+requiredEditorServicesVersion\s+=\s+"(.*)"'
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply tools/repoUpdateScripts/updateExtensionVersions.ps1]: I doubt it too, but while it's in the code I don't want to let it rot

File: tools/repoUpdateScripts/updateExtensionVersions.ps1 [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        $Version
+    )
+
+    $vstsDockerFileContent = Get-Content -Raw $DockerFilePath
+    $vstsDockerFileVersionSpan = FindVstsBuildVersionSpan -DockerFileContent $vstsDockerFileContent
+    $newDockerFileContent = ReplaceStringSegment -String $vstsDockerFileContent -NewSegment $Version -StartIndex $vstsDockerFileVersionSpan.Start -EndIndex $vstsDockerFileVersionSpan.End
+    SetFileContent -FilePath $DockerFilePath -Value $newDockerFileContent
+}
+
+function GetMarketplaceVersionFromSemVer
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply tools/repoUpdateScripts/updateExtensionVersions.ps1]: No, it converts a SemVer v2 version to what we've been using in the marketplace

File: tools/FileUpdateTools.psm1 [hunk 9]
```diff
@@ -0,0 +1,245 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply tools/FileUpdateTools.psm1]: It seemed like unnecessary work for an internal module, but since you ask I will

File: tools/GitHubTools.psm1 (hunk 4 above)
Comment (rjmholt) [reply tools/GitHubTools.psm1]: https://stackoverflow.com/questions/13839879/how-to-determine-the-humanish-part-of-a-git-repository

Looking into this

Comment (rjmholt) [reply tools/GitHubTools.psm1]: Ok I have a thing that works I think

File: tools/postReleaseScripts/updateAzureDataStudio.ps1 [hunk 10]
```diff
@@ -218,33 +218,23 @@ function UpdateGalleryFile
...
+    $startOffset = Get-StringOffsetFromSpan -String $galleryFileContent -EndLine $span.Start.Line -Column $span.Start.Column
+    $endOffset = Get-StringOffsetFromSpan -String $galleryFileContent -EndLine $span.End.Line -StartLine $span.Start.Line -Column $span.End.Column -InitialOffset $startOffset
 
     # Create the new file contents with the inserted segment
-    $newGalleryFileContent = ReplaceStringSegment -String $galleryFileContent -NewSegment $entryStr -StartIndex $startOffset -EndIndex ($endOffset+1) -AutoIndent
+    $newGalleryFileContent = New-StringWithSegment -String $galleryFileContent -NewSegment $entryStr -StartIndex $startOffset -EndIndex ($endOffset+1) -AutoIndent
 
     # Write out the new entry
-    SetFileContent $GalleryFilePath $newGalleryFileContent
+    Set-Content -Path $GalleryFilePath -Value $newGalleryFileContent -Encoding utf8NoBOM -NoNewline
```
Comment (TylerLeonhardt) [inline tools/postReleaseScripts/updateAzureDataStudio.ps1]: I think this means it'll only work on PS6 (`utf8noBOM`) - which I'm ok with for this script, but wanted to call this out

File: tools/postReleaseScripts/updateAzureDataStudio.ps1 [hunk 11]
```diff
@@ -271,4 +261,4 @@ $prParams = @{
     GitHubToken = $GitHubToken
     FromOrg = 'rjmholt'
 }
-OpenGitHubPr @prParams
\ No newline at end of file
+New-GitHubPR @prParams
```
Comment (TylerLeonhardt) [inline tools/postReleaseScripts/updateAzureDataStudio.ps1]: nit, newline

File: tools/postReleaseScripts/updateExtensionVersions.ps1 [hunk 12]
```diff
@@ -181,8 +183,8 @@ function UpdateDockerFileVersion
 
     $vstsDockerFileContent = Get-Content -Raw $DockerFilePath
     $vstsDockerFileVersionSpan = FindVstsBuildVersionSpan -DockerFileContent $vstsDockerFileContent
-    $newDockerFileContent = ReplaceStringSegment -String $vstsDockerFileContent -NewSegment $Version -StartIndex $vstsDockerFileVersionSpan.Start -EndIndex $vstsDockerFileVersionSpan.End
-    SetFileContent -FilePath $DockerFilePath -Value $newDockerFileContent
+    $newDockerFileContent = New-StringWithSegment -String $vstsDockerFileContent -NewSegment $Version -StartIndex $vstsDockerFileVersionSpan.Start -EndIndex $vstsDockerFileVersionSpan.End
+    Set-Content -Path $DockerFilePath -Value $newDockerFileContent -Encoding utf8NoBOM -NoNewline
```
Comment (TylerLeonhardt) [inline tools/postReleaseScripts/updateExtensionVersions.ps1]: Same comment on "requiring PS6"

File: tools/postReleaseScripts/updateExtensionVersions.ps1 [hunk 13]
```diff
@@ -259,25 +260,25 @@ UpdateDockerFileVersion -DockerFilePath $paths.vstsDockerFile -Version $marketPl
...
     Branch = $branchName
-    Title = "Update version to v$newVersionStr"
+    Title = "Update version to v$NewVersion"
     Description = $PRDescription
     GitHubToken = $GitHubToken
     FromOrg = 'rjmholt'
 }
-OpenGitHubPr @prParams
\ No newline at end of file
+New-GitHubPR @prParams
```
Comment (TylerLeonhardt) [inline tools/postReleaseScripts/updateExtensionVersions.ps1]: nit, newline

File: tools/postReleaseScripts/updateAzureDataStudio.ps1 (hunk 10 above)
Comment (rjmholt) [reply tools/postReleaseScripts/updateAzureDataStudio.ps1]: Yeah, that's right

Comment (rjmholt) [reply tools/postReleaseScripts/updateAzureDataStudio.ps1]: Basically the semver handling means this will only work in PS 6, so with that in mind I've gone with a few PS6-specific things, since we can always opt for pwsh in AzDO (and unlike building, it's not important to be able to run release code in WinPS-only contexts)

File: tools/repoUpdateScripts/updateAzureDataStudio.ps1 (hunk 2 above)
Comment (rjmholt) [reply tools/repoUpdateScripts/updateAzureDataStudio.ps1]: Turns out there was complicated file-reading logic that needed to be done when `$NewVersion` wasn't provided, so I had to move this back out

File: tools/postReleaseScripts/updateAzureDataStudio.ps1 (hunk 10 above)
Comment (TylerLeonhardt) [reply tools/postReleaseScripts/updateAzureDataStudio.ps1]: Should probably add this then:

```
#Requires -Version 6
```
